Chunkify time panel data density graphs

jprochazk opened the pull request:
### What

Part 1 of:
- https://github.com/rerun-io/rerun/issues/6065

This adds a new option, `Chunk-based data density graph`, which when enabled switches the timeline from rendering using `TimeHistogram` to rendering chunks directly, using range queries in the UI code.

To test, enable `Options -> Chunk-based data density graph`

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/6847?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/6847?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!

- [PR Build Summary](https://build.rerun.io/pr/6847)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

emilk commented on crates/viewer/re_time_panel/src/data_density_graph.rs:
```diff
@@ -366,6 +367,206 @@ fn smooth(density: &[f32]) -> Vec<f32> {
+            add_data_point(chunk_timeline.time_range(), events);
+        }
+    } else if let Some(components) = store.all_components(&timeline, &item.entity_path) {
+        for component_name in components {
+            let chunks =
+                db.store()
+                    .range_relevant_chunks(&query, &item.entity_path, component_name);
```
This will likely return the same chunks many times, e.g. if you log a point cloud with both color and positions, it will go into the same chunk, so we need to deduplicate the chunks. Probably nicest to do so in some helper functions in `ChunkStore` so we just call `range_relevant_chunks_for_entity` here.

emilk commented on crates/viewer/re_time_panel/src/data_density_graph.rs:
```diff
@@ -366,6 +367,206 @@ fn smooth(density: &[f32]) -> Vec<f32> {
+
+            add_data_point(chunk_timeline.time_range(), events);
+        }
+    } else if let Some(components) = store.all_components(&timeline, &item.entity_path) {
```
We also need to recursively do this for all child entities

teh-cmc commented on crates/store/re_chunk/src/chunk.rs:
```diff
@@ -218,6 +218,16 @@ impl Chunk {
             .collect()
     }
 
+    #[inline]
+    pub fn num_events(&self) -> usize {
+        self.components.values().map(|array| array.len()).sum()
+    }
+
+    #[inline]
+    pub fn num_events_for_component(&self, component_name: ComponentName) -> Option<usize> {
+        self.components.get(&component_name).map(|v| v.len())
+    }
+
```
3 issues here:
* `num_events` is misleading, as a caller you'd think this is an alias for `Self::num_rows`.
* You cannot rely on the length of the underlying `ListArray` since component columns are sparse.
* Precise naming of arrays is super important in `re_chunk` because everything is some kind of arrow array in here; it quickly gets unintelligible.

```suggestion
    /// The cumulative number of events in this chunk.
    ///
    /// I.e. how many _component batches_ ("cells") were logged in total?
    //
    // TODO(cmc): This needs to be stored in chunk metadata and transported across IPC.
    #[inline]
    pub fn num_events_cumulative(&self) -> usize {
        // Reminder: component columns are sparse, we must take a look at the validity bitmaps.
        self.components
            .values()
            .map(|list_array| {
                list_array.validity().map_or_else(
                    || list_array.len(),
                    |validity| validity.len() - validity.unset_bits(),
                )
            })
            .sum()
    }

    /// The number of events in this chunk for the specified component.
    ///
    /// I.e. how many _component batches_ ("cells") were logged in total for this component?
    //
    // TODO(cmc): This needs to be stored in chunk metadata and transported across IPC.
    #[inline]
    pub fn num_events_for_component(&self, component_name: ComponentName) -> Option<usize> {
        // Reminder: component columns are sparse, we must check validity bitmap.
        self.components.get(&component_name).map(|list_array| {
            list_array.validity().map_or_else(
                || list_array.len(),
                |validity| validity.len() - validity.unset_bits(),
            )
        })
    }

emilk commented on crates/viewer/re_time_panel/src/data_density_graph.rs:
```diff
@@ -366,6 +374,253 @@ fn smooth(density: &[f32]) -> Vec<f32> {
+        add_data_point(chunk, time_range, num_events);
+    }
+
+    let hovered_x_range = (time_ranges_ui
+        .x_from_time_f32(hovered_time_range.min().into())
+        .unwrap_or(f32::MAX)
+        - MARGIN_X)
+        ..=(time_ranges_ui
+            .x_from_time_f32(hovered_time_range.max().into())
+            .unwrap_or(f32::MIN)
+            + MARGIN_X);
```
nit: could use `Rangef` instead of Rust's annoying ranges

emilk commented on crates/viewer/re_time_panel/src/data_density_graph.rs:
```diff
@@ -366,6 +374,253 @@ fn smooth(density: &[f32]) -> Vec<f32> {
+        };
+        visitor(Arc::clone(&chunk), chunk_timeline, num_events);
+    } else {
+        for component_name in chunk.component_names() {
+            let chunk = Arc::new(chunk.range(&query, component_name));
+            let Some(num_events) = chunk.num_events_for_component(component_name) else {
+                continue;
+            };
+            let Some(chunk_timeline) = chunk.timelines().get(&timeline) else {
+                continue;
+            };
+            visitor(Arc::clone(&chunk), chunk_timeline, num_events);
+        }
```
I think this should be turned into a `chunk.range_entity(&query)` or something, that also de-duplicates the cunks.

Maybe @teh-cmc has some thoughts here

emilk replied:
Actually, I think we want to use the query cache… once it is fully built :)

emilk replied:
Actually no - we change the view too often for a cache to be helpful.

I think a better approach is this:
- if it is a small chunk, do a linear scan
- if it is a big chunk that is sorted on the timeline, to a cheap binary-search
- if it is a big chunk that is unsorted, then return the whole chunk (bad picking behavior, but performance is more important)

Hopefully we shouldn't have large unsorted chunks, and the compaction code should make sure of that, because large unsorted chunks will always be bad for performance

emilk commented on crates/viewer/re_time_panel/src/data_density_graph.rs:
```diff
@@ -366,6 +374,253 @@ fn smooth(density: &[f32]) -> Vec<f32> {
+    }
+}
+
+fn visit_relevant_chunks(
```
Needs docstring: does it deduplicate?

emilk commented on crates/viewer/re_time_panel/src/data_density_graph.rs:
```diff
@@ -366,6 +374,253 @@ fn smooth(density: &[f32]) -> Vec<f32> {
+    component_name: Option<ComponentName>,
+    timeline: Timeline,
+    time_range: ResolvedTimeRange,
+    mut visitor: impl FnMut(Arc<Chunk>, &ChunkTimeline, usize),
```
```suggestion
    mut visitor: impl FnMut(Arc<Chunk>, &TimeRange, usize),
```

emilk commented on crates/viewer/re_time_panel/src/data_density_graph.rs:
```diff
@@ -366,6 +374,253 @@ fn smooth(density: &[f32]) -> Vec<f32> {
+        };
+        visitor(Arc::clone(&chunk), chunk_timeline, num_events);
+    } else {
+        for component_name in chunk.component_names() {
+            let chunk = Arc::new(chunk.range(&query, component_name));
+            let Some(num_events) = chunk.num_events_for_component(component_name) else {
+                continue;
+            };
+            let Some(chunk_timeline) = chunk.timelines().get(&timeline) else {
+                continue;
+            };
+            visitor(Arc::clone(&chunk), chunk_timeline, num_events);
+        }
```
```suggestion
    let chunk = Arc::new(chunk.range_all(&query));
    visitor(Arc::clone(&chunk), chunk_timeline, num_events);
```

@teh-cmc ptal

emilk commented on crates/viewer/re_time_panel/src/data_density_graph.rs:
```diff
@@ -366,6 +374,253 @@ fn smooth(density: &[f32]) -> Vec<f32> {
+                    return;
+                };
+
+                for component_name in components {
+                    let chunks =
+                        db.store()
+                            .range_relevant_chunks(&query, entity_path, component_name);
+
+                    for chunk in chunks {
+                        let Some(chunk_timeline) = chunk.timelines().get(&timeline) else {
+                            continue;
+                        };
+
+                        if seen.contains(&chunk.id()) {
+                            continue;
+                        }
+                        seen.insert(chunk.id());
+
+                        visitor(
+                            Arc::clone(&chunk),
+                            chunk_timeline,
+                            chunk.num_events_cumulative(),
+                        );
+                    }
+                }
```
We should have an API for finding the relevant chunks for a given entity, which would save a lot of CPU here.

@teh-cmc do you agree?

```suggestion
                let chunks =
                    db.store()
                        .range_relevant_chunks_for_entity(&query, entity_path);

                for chunk in chunks {
                    let Some(chunk_timeline) = chunk.timelines().get(&timeline) else {
                        continue;
                    };

                    visitor(
                        Arc::clone(&chunk),
                        chunk_timeline,
                        chunk.num_events_cumulative(),
                    );
                }
```

emilk commented on crates/viewer/re_time_panel/src/data_density_graph.rs:
```diff
@@ -366,6 +374,253 @@ fn smooth(density: &[f32]) -> Vec<f32> {
+    } else {
+        let mut seen = HashSet::new();
+        if let Some(subtree) = db.tree().subtree(entity_path) {
+            subtree.visit_children_recursively(&mut |entity_path, _| {
```
does this include self? if so maybe we should give it a better name 🤔

emilk commented on crates/viewer/re_ui/src/command.rs:
```diff
@@ -346,6 +353,8 @@ impl UICommand {
             #[cfg(target_arch = "wasm32 ")]
             Self::ViewportMode(_) => None,
+
+            Self::ToggleChunkBasedDataDensityGraph => Some(ctrl_shift(Key::D)),
```
```suggestion
            Self::ToggleChunkBasedDataDensityGraph => Some(cmd_shift(Key::D)),
```

more mac friendly :)

emilk reviewed (approved):
Looks like a good first step! But I think we should consider entity-level APIs for Chunks to make this code less awkward and more performant

emilk commented:
I think it would be great with some test code for this, testing different scenarios:
* A lot of tiny chunks
* A few huge, sorted chunks
* A few huge, unsorted chunks
* An examples with gaps

teh-cmc commented:
I'm not sure I understand the UX here.

I open https://rerun.io/viewer/pr/6847?url=https%3A%2F%2Fapp.rerun.io%2Fversion%2Fnightly%2Fexamples%2Fnuscenes_dataset.rrd and then hover randomly over the time panel and I cannot quite make sense of the behaviour. Sometimes I get an hover pop-up, sometimes not, I'm not sure why?
I expect to always have a hover pop-up no matter what, if only to tell me that I can only see very coarse data because <reasons>.


https://github.com/rerun-io/rerun/assets/2910679/a8ab7ffa-a169-4283-b125-d551bcd4e8a7

Am I missing something?

teh-cmc commented on crates/viewer/re_time_panel/src/data_density_graph.rs:
```diff
@@ -366,6 +374,253 @@ fn smooth(density: &[f32]) -> Vec<f32> {
+    }
+}
+
+// if sorted:
+// - binary search, show hovered things
+// - if its small enough, full scan
+// - otherwise, hover full chunk
```
Is that a TODO or something?

jprochazk replied:
Leftover from discussion with Emil, resulting in https://github.com/rerun-io/rerun/pull/6847#discussion_r1673948501

emilk commented:
> I'm not sure I understand the UX here.
> 
> I open [rerun.io/viewer/pr/6847?url=https://app.rerun.io/version/nightly/examples/nuscenes_dataset.rrd](https://rerun.io/viewer/pr/6847?url=https%3A%2F%2Fapp.rerun.io%2Fversion%2Fnightly%2Fexamples%2Fnuscenes_dataset.rrd) and then hover randomly over the time panel and I cannot quite make sense of the behaviour. Sometimes I get an hover pop-up, sometimes not, I'm not sure why? I expect to always have a hover pop-up no matter what, if only to tell me that I can only see very coarse data because <reasons>.

The reason, I think, is because the current code always renders a chunk as one solid block. This means you can hover it and in some parts of the block there will actually be data behind it, and in other places there will be not. It kind of looks like this:


```
[o oo  o ooo o o o           o                     o o oo  o      o]
```

where `o` is an actual event (invisible) and `[…]` represents the chunk containing it.

When hovered, the code does a smaller range-query to see if there is actually something there, and sometimes there are, sometimes there isn't.

That explains the why, but we for sure should come up with a better UX.

First of all: the plan is to only summarize huge chunks as blocks, or small chunks if there are too many of them, but there will always be times when we have this problem.

I think you are right that we should always show _something_, at least what entity and component you hover. That we can fix right away.

Currently when you hover exactly one event, you see its content. If you hover multiple events you only see "Hovering 13 events", which kind of sucks.

---

I wonder if we should just switch out the range-query we do on hover for a latest-at query. That way hovering is sort of like a preview to time-scrubbing: this is the data you'd see if you moved the time cursor here. We could then also highlight where that event actually is in the timeline, which could be one pixel left of the mouse cursor, or twenty.

Of course, this brings up other questions: what data do we show when we hover the data density graph of an ancestor entity - do we show the latest-at value for all components for all decedents?

emilk commented on crates/viewer/re_time_panel/src/data_density_graph.rs:
```diff
@@ -366,6 +374,304 @@ fn smooth(density: &[f32]) -> Vec<f32> {
+    timeline: Timeline,
+    time_range: ResolvedTimeRange,
+    mut visitor: impl FnMut(Arc<Chunk>, ResolvedTimeRange, usize),
+) {
```
```suggestion
) {
    re_tracing::profile_function!();
```

jprochazk commented:
Updated after design call with Katya and @emilk:

- The "timeline cursor preview" that used to only show when hovering over the timeline rect now shows when hovering anywhere in the time area
- Set the "timeline cursor preview" to use a brighter color, it should be more visible now
- Clicking anywhere in the time area now takes you to that point on the timeline (as the preview suggests)
- Removed the range highlight
- Removed `num_events` from the tooltip

For testing, I added `tests/rust/test_data_density_graph`, based on https://github.com/rerun-io/rerun/pull/6847#issuecomment-2222843512

emilk commented:
Double-clicking to reset the pan/zoom of the streams panel view no longer works

emilk commented:
When hovering something in a chunk I want to know when in time the thing I hover is at.

![image](https://github.com/user-attachments/assets/3ac08849-061a-4d80-a6ed-25aa43a68844)

I think the individual time-point of the latest-at query should be shown in the tooltip, but also visually.

For the visually part I think either that's where the time-cursor snaps to on hover, or we add a circle or something to highlight when that time-point is.

Both these can wait for a later PR though, but definitely before we switch to the new chunk design.

jprochazk commented:
> Double-clicking to reset the pan/zoom of the streams panel view no longer works

Should be working now. This feels like something that should eventually be added to the release checklist
